### PR TITLE
Add example for configuring JS collocation in _Layout.cshtml

### PR DESCRIPTION
**Pull request by sliekens**
This advice should work for Razor pages and for MVC views. I'm not sure about scripts provided by Razor class libraries, I don't use them personally.

***EDIT:***
---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: c90104e1-65b1-0007-e56d-0dffef4b1a44
* Version Independent ID: 5c95e5dc-dc10-8658-91fd-1d55d3959c37
* Content: [Script Tag Helper in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/script-tag-helper?branch=pr-en-us-28802&view=aspnetcore-7.0#asp-src)
* Content Source: [aspnetcore/mvc/views/tag-helpers/built-in/script-tag-helper.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/mvc/views/tag-helpers/built-in/script-tag-helper.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/mvc/views/tag-helpers/built-in/samples/ScriptTagHelper/wwwroot/lib/jquery-validation/LICENSE.md](https://github.com/dotnet/AspNetCore.Docs/blob/80b794a5ad8544ea328beb9e6a4bf95104ff3f06/aspnetcore/mvc/views/tag-helpers/built-in/samples/ScriptTagHelper/wwwroot/lib/jquery-validation/LICENSE.md) | [aspnetcore/mvc/views/tag-helpers/built-in/samples/ScriptTagHelper/wwwroot/lib/jquery-validation/LICENSE](https://review.learn.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/samples/ScriptTagHelper/wwwroot/lib/jquery-validation/LICENSE?branch=pr-en-us-33190) |

### CODE REVIEW

**Comment by guardrex** [comment]
Hello @sliekens ... It's a reasonable tip, but this would need to call out that every page of the layout must have a collocated JS file.

@Rick-Anderson ... Are you interested in this?

**Comment by sliekens** [comment]
@guardrex I see why you might think that it always adds a script tag to every page, but `asp-src-include` omits the script tag when there are no matching files. So, this allows a mix of pages with and without collocated scripts.

**Comment by guardrex** [comment]
I see. I looked it up, and it didn't say that was the case ...

https://learn.microsoft.com/en-us/aspnet/core/mvc/views/tag-helpers/built-in/script-tag-helper?view=aspnetcore-8.0#asp-src-include

**Comment by sliekens** [comment]
For context about why this works:

``` html
<!-- <script asp-src-include="@(ViewContext.View.Path).js"></script> -->
<script asp-src-include="/Pages/Index.cshtml.js"></script>
```

The ScriptTagHelper sees there is no `src` attribute, so it deletes the original `<script>` tag.

https://github.com/dotnet/aspnetcore/blob/48a07213d4d8df15c6dffccd161844842c196998/src/Mvc/Mvc.TagHelpers/src/ScriptTagHelper.cs#L307-L313

Then it uses the string `"/Pages/Index.cshtml.js"` as a globbing pattern to find matching files. It then renders a new script tag for each matching file. Since the pattern doesn't include any wildcards (`*`), it should either find 1 or 0 matching files. If no matching file is found, no script tag is rendered.

**Comment by guardrex** [comment]
> update the PR with a note about this behavior?

Probably best to wait for Rick to see this. It seems harmless enough as a tip. Rick might want to document the underlying behavior in the Script Tag Helper article and cross-link to it from here. The cross-link will probably be to `xref:mvc/views/tag-helpers/built-in/script-tag-helper#asp-src-include`. ***And*** ... he can dissolve the INCLUDE file at the same time.

He'll be along in a few hours ... he's in a different time zone 🌔.

File: aspnetcore/includes/js-collocation.md [hunk 1]
```diff
@@ -26,6 +26,12 @@ Collocated JS files are publicly addressable using the ***path to the file in th
   }
   ```
 
+  You can configure the default layout `(Pages/Shared/_Layout.cshtml)` to always include collocated JS files, eliminating the need to configure each page individually:
```
Comment (Rick-Anderson) [inline aspnetcore/includes/js-collocation.md:29]:
```suggestion
The default layout `(Pages/Shared/_Layout.cshtml)` can be configured to include collocated JS files, eliminating the need to configure each page individually:
```

**Review by Rick-Anderson — CHANGES_REQUESTED** [review]

**Comment by sliekens** [comment]
@Rick-Anderson sorry, what's an RP sample? 😀

**Comment by Rick-Anderson** [comment]
> @Rick-Anderson sorry, what's an RP sample? 😀
> 
> /edit: having just realized RP means Razor Pages, what are your expectations for such a sample?

Sorry for using RP, I should have spelled it out. Just the simplest possible sample that demonstrates JS collocation in `_Layout.cshtml` and `asp-src-include` omits the script tag when there are no matching files. So, this allows a mix of pages with and without collocated scripts. 

You can do the sample in another PR and add add closes ` #33400`

**Review by Rick-Anderson — APPROVED** [review]

**Comment by sliekens** [comment]
Okay @Rick-Anderson I canceled the automerge because I had already pushed a part of the sample to this branch. I will move the changes over to a new PR.

**Comment by Rick-Anderson** [comment]
@sliekens sorry about that, merging here would have been fine. I didn't realize you could generate a sample that fast and didn't want to hold up this PR. Either way is fine.

**Comment by sliekens** [comment]
Okay, I pushed the sample to this branch after all.

I used `dotnet new razor`, removed the `Privacy` page and instead added two pages.

1. a page with a collocated JS script
   image
2. another page without any script
   image

**Comment by sliekens** [comment]
@Rick-Anderson if you could review again, I hope this is what you had in mind.

**Comment by Rick-Anderson** [comment]
> @Rick-Anderson if you could review again, I hope this is what you had in mind.

Looks great. I'll test it tomorrow.

File: aspnetcore/includes/js-collocation.md [hunk 2]
```diff
@@ -26,6 +26,12 @@ Collocated JS files are publicly addressable using the ***path to the file in th
   }
   ```
 
+The default layout `(Pages/Shared/_Layout.cshtml)` can be configured to include collocated JS files, eliminating the need to configure each page individually:
```
Comment (Rick-Anderson) [inline aspnetcore/includes/js-collocation.md:29]:
```suggestion
The default layout `Pages/Shared/_Layout.cshtml` can be configured to include collocated JS files, eliminating the need to configure each page individually:
```

**Comment by Rick-Anderson** [comment]
@sliekens  moved to #33697, I carried your commit so you'll get credit for the merge.

**Comment by sliekens** [comment]
Cool, thanks 😎👍

**Pull request closed without merging**